### PR TITLE
Matchmaking

### PR DESCRIPTION
**Pull request by hferraud**
To test the matchmaking.

Launch the client
```bash
cd matchmaking/client
npm install
node server.js
```
Launch the server:
```bash
cd matchmaking
python -m src.matchmaking
```
Now go to http://localhost:5000

### CODE REVIEW

File: matchmaking/src/authenticate.py [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+        )
+        return payload
+    except Exception:
+        return None
+
+
+def get_user(user_id: int, token: str) -> Optional[dict]:
+    headers = {'Authorization': token}
+    try:
+        response = requests.get(f'http://user-management-nginx/user/{user_id}/', headers=headers)
```
Comment (tdameros) [inline matchmaking/src/authenticate.py:37]: Maybe use URL variable for every request on user-management ?

**Review by tdameros — CHANGES_REQUESTED** [review]

File: matchmaking/src/authenticate.py (hunk 1 above)
Comment (hferraud) [reply matchmaking/src/authenticate.py]: 7f92e5c

Moved user-management in settings, we should use common folder in the future, see issue https://github.com/tdameros/42-transcendence/issues/34

File: matchmaking/src/authenticate.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+def authenticate_user(token: str) -> Optional[dict]:
+    payload = decode_jwt(token)
+    if payload is None:
+        return None
+    user_id = payload.get('user_id')
+    if user_id is None:
+        return None
+    # TODO: Uncomment the following lines when the `user/` endpoint is implemented
+    # user = get_user(user_id, token)
+    # return user
```
Comment (EthanDelage) [inline matchmaking/src/authenticate.py:18]: delete this

File: matchmaking/src/authenticate.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        )
+        return payload
+    except Exception:
+        return None
+
+
+def get_user(user_id: int, token: str) -> Optional[dict]:
+    headers = {'Authorization': token}
+    try:
+        response = requests.get(f'{settings.USER_MANAGEMENT_USER_ENDPOINT}{user_id}', headers=headers)
```
Comment (EthanDelage) [inline matchmaking/src/authenticate.py:37]: In my opinion, it's better to write the request url like this: `{settings.USER_MANAGEMENT_URL}user/{user_id}`.
This allows you to see on which endpoint the request is made.

Comment (EthanDelage) [inline matchmaking/src/authenticate.py:37]: Just try this line

**Review by EthanDelage — CHANGES_REQUESTED** [review]

File: matchmaking/src/authenticate.py (hunk 3 above)
Comment (hferraud) [reply matchmaking/src/authenticate.py:37]: It's debatable, if the `user/` endpoint change in the future we will need to change it manually in the code

Comment (hferraud) [reply matchmaking/src/authenticate.py:35]: 85ee9e3

File: matchmaking/src/authenticate.py (hunk 2 above)
Comment (hferraud) [reply matchmaking/src/authenticate.py]: 85ee9e3

File: matchmaking/src/authenticate.py (hunk 3 above)
Comment (hferraud) [reply matchmaking/src/authenticate.py:35]: It will be further discussed when solving issue https://github.com/tdameros/42-transcendence/issues/34

**Review by EthanDelage — APPROVED** [review]

**Review by V-Fries — APPROVED** [review]

File: matchmaking/src/matchmaking.py [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+    def disconnect(self, sid):
+        for user in self.queue:
+            if user.get('sid') == sid:
+                self.queue.remove(user)
+
+    def queue_join(self, sid, data):
+        user_id = data.get('user_id')
+        elo = data.get('elo')
+        if elo is None or user_id is None:
+            self.sio.disconnect(sid)
```
Comment (aLeuleu) [inline matchmaking/src/matchmaking.py:113]: TODO : send an error

File: matchmaking/test/test.py
Comment (aLeuleu) [inline matchmaking/test/test.py:1]: To remember : let's try to test all of these features on a django module somewhere else

**Review by aLeuleu — APPROVED** [review]

File: matchmaking/src/matchmaking.py (hunk 4 above)
Comment (hferraud) [reply matchmaking/src/matchmaking.py:113]: issue created https://github.com/tdameros/42-transcendence/issues/50

File: matchmaking/test/test.py
Comment (hferraud) [reply matchmaking/test/test.py:1]: issue created https://github.com/tdameros/42-transcendence/issues/51

**Pull request merged**